src/dataflow/decode/mod.rs

```diff
@@ -118,6 +118,306 @@ where
+                interchange::avro::EnvelopeType::Upsert,
+            ),
+            &op_name,
+        ),
+        (DataEncoding::Text, DataEncoding::Avro(val_enc)) => decode_kafka_upsert_inner(
+            stream,
+            OffsetDecoderState {
+                datum_func: text_to_datum,
+            },
+            avro::AvroDecoderState::new(
+                &val_enc.value_schema,
+                val_enc.schema_registry_url,
+                interchange::avro::EnvelopeType::Upsert,
+            ),
+            &op_name,
+        ),
+        (DataEncoding::Avro(key_enc), DataEncoding::Avro(val_enc)) => decode_kafka_upsert_inner(
+            stream,
+            avro::AvroDecoderState::new(
+                &key_enc.value_schema,
+                key_enc.schema_registry_url,
+                interchange::avro::EnvelopeType::None,
+            ),
+            avro::AvroDecoderState::new(
+                &val_enc.value_schema,
+                val_enc.schema_registry_url,
+                interchange::avro::EnvelopeType::Upsert,
+            ),
+            &op_name,
+        ),
+        (DataEncoding::Text, DataEncoding::Bytes) => decode_kafka_upsert_inner(
+            stream,
+            OffsetDecoderState {
+                datum_func: text_to_datum,
+            },
+            OffsetDecoderState {
+                datum_func: bytes_to_datum,
+            },
+            &op_name,
+        ),
+        (DataEncoding::Bytes, DataEncoding::Bytes) => decode_kafka_upsert_inner(
+            stream,
+            OffsetDecoderState {
+                datum_func: bytes_to_datum,
+            },
+            OffsetDecoderState {
+                datum_func: bytes_to_datum,
+            },
+            &op_name,
+        ),
+        (DataEncoding::Text, DataEncoding::Text) => decode_kafka_upsert_inner(
+            stream,
+            OffsetDecoderState {
+                datum_func: text_to_datum,
+            },
+            OffsetDecoderState {
+                datum_func: text_to_datum,
+            },
+            &op_name,
+        ),
+        _ => unreachable!(),
+    };
+    decoded_stream.unary(Pipeline, name, move |_, _| {
+        move |input, output| {
+            input.for_each(|cap, data| {
+                let mut session = output.session(&cap);
+                let mut v = Vec::new();
+                data.swap(&mut v);
+                session.give_iterator(v.into_iter().map(|(key, value, timestamp)| {
+                    if let Some(value) = value {
+                        let mut value_with_key = RowPacker::new();
+                        value_with_key.extend_by_row(&key);
+                        value_with_key.extend_by_row(&value);
+                        (key, Some(value_with_key.finish()), timestamp)
+                    } else {
+                        (key, None, timestamp)
+                    }
+                }));
+            });
+        }
+    })
+}
+
+pub fn kafka_value_source<G>(
```

Comment:
Can this function be replaced using https://docs.rs/timely/0.11.1/timely/dataflow/operators/map/trait.Map.html ?

`let source = source.map(|((_key, payload), aux_num)| (payload, aux_num))`